### DISCUSSION TITLE
Matcher的超时提示，以及自定义超时时间

### DISCUSSION THREAD

**Discussion by PrunusSerrulata** [nonebot/discussions#474] [想法]
RT，举例来说我为了一个交互式游戏写了如下代码：
```python
matcher = on_command(...)

@matcher.handle()
async def _(...):
    # ...
    if ai_go_first: # 本局游戏AI先手
        state["ai_turn"] = True
        matcher.set_arg("cmd", Message())
    # ...

@matcher.got("cmd")
async def _(cmd: str = ArgPlainText(), ...):
    # ... 
    if not cmd: # 非开局随机到AI先手的情况
        # 处理玩家操作
    # ...
    while state["ai_turn"]: # AI操作
        # ...
        state["ai_turn"] = False # AI操作结束
    # ...
    if game_end:
        await matcher.finish("something")
    else:
        await matcher.reject("something") # 等待玩家再次输入
```
我希望在读取用户输入超时之后发送一条超时提醒。我可以使用hook在`run_postprocessor`里注册回调，然后在`run_preprocessor`里取消（此时已经读到用户的输入，取消之后`run_postprocessor`会注册新的回调），但`run_postprocessor`没法得到`PausedException`，`RejectedException`和`FinishedException`，不知道游戏是结束了还是继续等待用户输入（如果游戏结束就不需要再注册回调了）。虽然可以手动告知状态，但是这样写很麻烦，而且要写的地方多了也很容易漏掉或者出错。

此外，因为小游戏难度较高，我希望为它**单独**设置超时（例如普通命令2分钟，这个小游戏5分钟），并且在超时时能发送提醒（Nonebot本身不存在原生的超时提醒）。但是每次执行reject之后都会生成新的临时matcher，首先这个临时matcher的超时时长是写死的用户没法改，另外一点是我没有找到方法能够获取到它（我需要获取到它才能靠定时回调什么的来销毁它，但当它被触发之后我才能获取到，此时已经没有必要销毁了）。

标题里的两个需求实际上真正没法解决的是第二个，虽然我希望第一个也能得到解决（这样插件开发者就不用费心在很复杂的交互流程里操心跳转时的状态）。给用户自定义reject之类方法生成的临时matcher的超时时间的能力应该不难？

我接触插件开发还不久，如有不妥之处还请不吝赐教。

**Comment by yanyongyu** [comment]
第一点，你可以直接在handler里自行捕获异常，装饰器或者依赖注入都可以获得。

第二点，当前nb里的超时是lazy检测，所以没有提供回调方法。这里推荐使用 nonebot-plugin-waiter 插件，应该能满足你的需求。

**Reply by PrunusSerrulata** [reply]
好的，感谢帮助，我去看看
